### ISSUE TITLE
Intercept updates are way too large

### ISSUE THREAD

**Issue by etiennekintzler** [creme-ml/creme#332]
Hello !

Currently working on #270, I am doing trials on the NYC taxi dataset. I use the following gist for the pipeline : [HKML2020.ipynb](https://gist.github.com/MaxHalford/e23c4fe26c035b818bc40cbdde9c3a8f). Versions used : **`creme`** : 0.6.1 ,  **Python** : 3.7.6, **OS** : Ubuntu 18.04.5 LTS.

**The intercept updates are extreme for some input change**. More specifically, at observation 546, the feature `hour` change for the first time from 0 to 1 and the transformed feature from around 0 to about 33. From then one, the intercept update become extremely large (1e9 to 1e10) and oscillate between positive and negative. 

This problem appears for the default learning rate of `lr=0.1` but is still present for lower values such as 0.05, 0.025, 0.125, although to a lesser extent. The problem disappears for value of lr of 0.001 and lower.

The predictions becomes extremely large as well (partly because of intercept but probably also because of weights). Below are the graphs that summarize the problem:

![image](https://user-images.githubusercontent.com/17943873/90954552-75fc5880-e475-11ea-9ee8-0583b66feb9e.png)

Below is the code snippet to reproduce the problem (and generate the graphs):

```python
# Package imports
import pandas as pd
import matplotlib.pyplot as plt
from creme import (
    optim,
    datasets,
    preprocessing,
    linear_model,
    metrics,
    compose
    )
import math
import calendar
import numbers 

# Preprocessing functions
def distances(trip):
    lat_dist = trip['dropoff_latitude'] - trip['pickup_latitude']
    lon_dist = trip['dropoff_longitude'] - trip['pickup_longitude']
    return {
        'manhattan_distance': abs(lat_dist) + abs(lon_dist),
        'euclidean_distance': math.sqrt(lat_dist ** 2 + lon_dist ** 2)
    }


def datetime_info(trip):
    day_no = trip['pickup_datetime'].weekday()
    return {
        'hour': trip['pickup_datetime'].hour,
        'day': calendar.day_name[day_no]
    }

# Helpers to generate the results (df and graphs)
def generate_pipeline(lr=0.1):
    extract_features = compose.TransformerUnion(distances, datetime_info)

    cat = compose.SelectType(str) | preprocessing.OneHotEncoder()
    num = compose.SelectType(numbers.Number) | preprocessing.StandardScaler()

    preprocess = compose.TransformerUnion(cat, num)

    pipeline = compose.Pipeline(
        extract_features,
        preprocess
    )
    pipeline |= linear_model.LinearRegression(optimizer=optim.SGD(lr=lr))
    
    return pipeline

def generate_result_df(data, lr=0.1):
    pipeline = generate_pipeline(lr=lr)

    out = []
    for (x,y) in data.take(600):
        transformed_x = pipeline.transform_one(x)
        y_pred = pipeline.predict_one(x)
        pipeline.fit_one(x=x, y=y)
        out += [dict(y=y, y_pred=y_pred, x=transformed_x, intercept=pipeline.steps["LinearRegression"].intercept)]

    out_df = pd.DataFrame(out)
    
    # Splitting the dict into columns
    features_names = list(out_df.x[0].keys())
    out_df[features_names] = pd.json_normalize(out_df.x)
    out_df.drop(columns="x", inplace=True)
    
    # Computing intercept increase
    out_df["intercept_increase"] = out_df.intercept.diff().fillna(0)

    return out_df

def generate_result_graphs(df, vertical_bar_index=546, vertical_bar_color="red", figsize=None):
    if figsize is None:
        figsize = (10, 10)

    fig, axes = plt.subplots(nrows=2, ncols=2, figsize=figsize)

    fig0 = df.hour.reset_index().plot.scatter(ax=axes[0, 0], x="index", y="hour", title="Hour (transformed)")
    fig0.axvline(x=vertical_bar_index, color=vertical_bar_color)

    fig1 = df.y_pred.plot(ax=axes[1, 0], title="Prediction")
    fig1.axvline(x=vertical_bar_index, color=vertical_bar_color)

    fig2 = df.intercept.plot(ax=axes[0, 1], title="Intercept")
    fig2.axvline(x=vertical_bar_index, color=vertical_bar_color)
    
    fig3 = df.intercept_increase.plot(ax=axes[1, 1], title="Intercept (variation)")
    fig3.axvline(x=vertical_bar_index, color=vertical_bar_color)
    

# Experiment
data = datasets.Taxis()
out_df = generate_result_df(data, lr=0.1)
generate_result_graphs(out_df)
```

**Comment by etiennekintzler** [comment]
> I reopen the issue to control the behaviour of the standard scaler with the feature hour

Ok great @raphaelsty ! I can rename the issue if necessary.

I will look into that as well, maybe do a toy example with a feature of 0 that becomes 1 after a long time in order to isolate the behavior.

> thanks for all these details @etiennekintzler. At least we have kept regression out of the problem. :-)

Yep :smiley:

**Comment by raphaelsty** [comment]
Hello @etiennekintzler,

To make sure that there are no problems when updating the standard scaler I did a test to compare the behavior of the Creme scaler standard and the Scikit Learn scaler standard. 

Both standard scalers provide the same results for the feature `hour` when using `std` and `mean` in the scaler:

```python
from creme import optim
from creme import datasets
from creme import preprocessing
from creme import optim
from creme import optim

import pandas as pd
import numpy  as np

import matplotlib.pyplot as plt

from sklearn import preprocessing as sk_preprocessing

# Creme 

data         = datasets.Taxis()
scaler       = preprocessing.StandardScaler()
hours_scaled_creme = []
hours        = []

for i, (x, y) in enumerate(data.take(600)):
    
    x = {'hour': x['pickup_datetime'].hour} 
    
    hours.append(x['hour'])
    
    x = scaler.fit_one(x).transform_one(x)
    
    hours_scaled_creme.append(x['hour'])


# Scikit Learn

# Default:
scaler = sk_preprocessing.StandardScaler()

# Without unit_variance
scaler_wuv = sk_preprocessing.StandardScaler(with_std = False)

hours_scaled_sklearn     = []
hours_scaled_sklearn_wuv = []

for hour in hours:
    
    hour = np.array([hour]).reshape(-1, 1)
    
    scaler = scaler.partial_fit(hour)
    
    scaler_wuv = scaler_wuv.partial_fit(hour)
        
    hours_scaled_sklearn.append(scaler.transform(hour)[0][0])
    
    hours_scaled_sklearn_wuv.append(scaler_wuv.transform(hour)[0][0])
    

df = pd.DataFrame(
    {
        'hours'      : hours, 
        'creme'      : hours_scaled_creme, 
        'sklearn'    : hours_scaled_sklearn, 
        'sklearn_wuv': hours_scaled_sklearn_wuv 
    }
)

assert df['sklearn'].round(6).equals(df['creme'].round(6))

plt.plot(range(600), df['creme'],       label = "hours_scaled_creme")
plt.plot(range(600), df['sklearn'],     label = "hours_scaled_sklearn")
plt.plot(range(600), df['sklearn_wuv'], label = "hours_scaled_sklearn without unit variance")
plt.plot(range(600), df['hours'],       label = "hour")
plt.legend()
plt.show()  
```

<img width="397" alt="image" src="https://user-images.githubusercontent.com/24591024/91081936-3174e680-e648-11ea-8fa2-02a0c3c50e64.png">

As you explained, removing the unit variance condition avoids the outlier observed in the output of the standard scaler. Besides Scikit Learn offers this [feature](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html) with the parameter `with_std` set to `True` by default.

I'll open an issue to discuss this feature. :-)
